commons/src/scanner/symbol/iterator/mod.rs

```diff
@@ -304,14 +349,54 @@ impl<'input> PeekingNext for SymbolIterator<'input> {
         Self: Sized,
         F: FnOnce(&Self::Item) -> bool,
     {
-        // Note: Not possible to restrict peek to return only symbols `next()` would return,
-        // because `peeking_next()` is needed in End- and PrefixMatcher.
-        // Using the same logic as in `next()` would result in endless loop inside `peeking_next()` => StackOverflow
+        if self.prefix_mismatch || self.end_reached() {
+            return None;
+        }
+
+        if !self.next_matching && !self.peek_matching {
+            if let Some(end_fn) = self.end_match.clone() {
+                let peek_index = self.peek_index();
+                self.peek_matching = true;
 
-        match &mut self.kind {
+                let end_matched = (end_fn)(self);
+
+                self.peek_matching = false;
```

Comment:
Ok, I think I get how these `*_matching` fields are used. Basically, that's the context for the matcher function to know whether we're in `peek` or `next` function call? 

If that's the case, it seems like it would be to forget changing these variables (always the case with mutating state). We could maybe use a different solution:

1. Change the signature of matcher function to take additional `MatcherContext` parameter: 
   * `MatcherContext` has functions or fields for checking which `*_matching` it is etc. 
   * It should be easy to create, maybe builder pattern
   * Benefit is, we cannot forget to create it, since function needs it as parameter
2. Do not implement `*Matcher` trait for `SymbolIterator` directly, rather wrap iterator in another struct that implements that trait 
   * this wrapper struct can have fields indicating what the context is
   * we implement `*Mathcer` trait for it
   * Benefit is, `SymbolIterator` cannot be passed into the matcher functions without being wrapped in matcher struct 
   * Might be a little harder to implement, maybe with generics? 
   * Might also be a little less flexible than using a second parameter for context, but maybe that's a good thing

Both of these options would probably require a little more boilerplate (another struct, implementation of the struct etc.), but I think it would improve code readability and also prevent some potential mistakes.

Reply:
I want to keep the `Matcher` traits as is, because these are used in all elements.
Adding complexity here would spread complexity to all element implementations.
With the internal flags, the complexity is bound to the two files in the `SymbolIterator` implementation. I think it is better to have this complexity here with the chance of getting it wrong, because it makes it easier on the element implementation side.

Reply:
Well, we could have both of those benefits by wrapping the iterator in a separate type that implements `Matcher` trait. This separate type could have a reference to the iterator internally, and additional needed context. We create it when we call the function, and it is dropped right after. That way you can't make a mistake by forgetting to reset state. 

It would be something like this: 

```rust
// instead of this 
self.peek_matching = true;
let end_matched = (end_fn)(self);
self.peek_matching = false;

// we would do something like
let end_matched = (end_fn)(Matcher { iter: self, state: MatcherState::Peek });

// where Matcher is something like: 
struct Matcher<'iter> {
    iter: &'iter mut SymbolIterator<'iter>, // or whatever necessary
    state: MatcherState, 
}

enum MatcherState {
    Peek,
    Next,
}
```

`Matcher` then just implements the `*Matcher` traits. The complexity would still be contained in iterator implementation and users of iterator and matching functions do not know about this at all, but the benefit is we do not have to keep track of state in iterator.

Reply:
I will create an issue for this once I fix the issue templates.